android/src/com/mapswithme/maps/MwmApplication.java

```diff
@@ -161,7 +161,43 @@ public String getApkPath()
     }
   }
 
-  public String getDataStoragePath()
+  private static boolean isMapStorage(String path)
+  {
+    final File dir = new File(path);
+    if (TextUtils.isEmpty(path) || !Utils.isDirectoryReadable(dir))
+      return false;
+
+    // Current map directory should contain folders with map indices at least, so use this fact as an indicator.
+    final File[] list = dir.listFiles(new FileFilter()
+    {
+      @Override
+      public boolean accept(File file)
+      {
+        return file.isDirectory();
+      }
+    });
+
+    return (list.length != 0);
+  }
+
+  private String findMapStoragePath()
+  {
+    String res = mPrefs.getString(Constants.STORAGE_PATH_PREF, "");
+    if (isMapStorage(res))
+      return res;
+
+    res = nativeGetString(Constants.STORAGE_PATH_PREF, "");
+    if (!isMapStorage(res))
```

Comment:
This logic is dangerous. Let's assume that user had valid storage path to external SD card. After this check, if external SD-card (or may be even internal one) is not available at this moment by any reason (ejected, still mounting, checked after device reboot, temporarily busy etc.) we forcedly replace it with a default one. If any maps or fonts resources or world files were downloaded there, they will be lost. And user will get a crash or should re-download maps again (what if he is in offline that moment?)

One of possible solutions I see is to display some UI dialog with explanation, that "Maps at <this path> are not available now. If you have ejected SD-card, please insert it back and press Try Again button. Otherwise your maps storage will be reset and you need to re-download maps again".
And two buttons in this dialog: Try Again (default one) and Reset Storage (dangerous button).
It would be great to constantly check storage availability, and if it becomes available again in a few seconds, then simply continue execution.